[WTF] Add overloads of WTF::arePointingToEqualData for non-nullable pointer-like types

weinig opened the pull request:
#### f7123211c62d1bfdaeb837b5f5c0b69827866e20
<pre>
[WTF] Add overloads of WTF::arePointingToEqualData for non-nullable pointer-like types
<a href="https://bugs.webkit.org/show_bug.cgi?id=294615">https://bugs.webkit.org/show_bug.cgi?id=294615</a>

Reviewed by Darin Adler.

Replaces the relatively common idiom of:

```
   foo.ptr() == bar.ptr() || foo.get() == bar.get();
```

used for Ref and UniqueRef that want value semantics, with
calls to an overload of WTF::arePointingToEqualData that
does the right thing.

* Source/WTF/wtf/Ref.h:
* Source/WTF/wtf/UniqueRef.h:
    - Adds overloads.

* Source/WebCore/css/CSSCustomPropertyValue.cpp:
* Source/WebCore/css/CSSFontStyleRangeValue.cpp:
* Source/WebCore/css/CSSVariableReferenceValue.cpp:
* Source/WebCore/css/calc/CSSCalcTree.h:
* Source/WebCore/css/values/CSSValueAggregates.h:
* Source/WebCore/css/values/color/CSSDynamicRangeLimitMix.h:
* Source/WebCore/platform/graphics/FontCascadeDescription.h:
* Source/WebCore/platform/graphics/avfoundation/objc/CDMSessionAVContentKeySession.mm:
* Source/WebCore/platform/graphics/filters/FELighting.cpp:
* Source/WebCore/platform/graphics/filters/FilterOperations.cpp:
* Source/WebCore/platform/graphics/transforms/TransformOperations.cpp:
* Source/WebCore/rendering/style/StyleCursorImage.cpp:
* Source/WebCore/rendering/style/StyleMultiImage.cpp:
* Source/WebCore/style/StyleCustomProperty.cpp:
* Source/WebCore/style/StyleCustomProperty.h:
* Source/WebCore/svg/graphics/filters/SVGFilter.cpp:
    - Adopt new arePointingToEqualData overloads.

Canonical link: <a href="https://commits.webkit.org/296358@main">https://commits.webkit.org/296358@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e27be1d74ecbd6fec22f037ac72d273154c389e9

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/108192 "2 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/27855 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/18275 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/113403 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/58678 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/110155 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/28555 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/36406 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/82152 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/111140 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/22626 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/97464 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/62583 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/22040 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/15600 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/58140 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/100781 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/91992 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/15656 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/116529 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/106753 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/35258 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/25964 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/91183 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/35632 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/93740 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/90978 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/23198 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/35862 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/13626 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/31039 "Build was cancelled. Recent messages:Printed configuration") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/35157 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/40713 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/131041 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/34891 "Built successfully") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/35570 "Found 2 new JSC stress test failures: wasm.yaml/wasm/stress/jitless-wasm-to-js-should-follow-to-wasm-call-frame-convention-2.js.default, wasm.yaml/wasm/stress/jitless-wasm-to-js-should-follow-to-wasm-call-frame-convention.js.default (failure)") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/38249 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/36558 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler reviewed (approved):
Looks good; perfect for clarity and likely good for performance. Do we think that the pointer comparisons added here are all valuable and not wasteful?

weinig committed:
[WTF] Add overloads of WTF::arePointingToEqualData for non-nullable pointer-like types
https://bugs.webkit.org/show_bug.cgi?id=294615

Reviewed by Darin Adler.

Replaces the relatively common idiom of:

```
   foo.ptr() == bar.ptr() || foo.get() == bar.get();
```

used for Ref and UniqueRef that want value semantics, with
calls to an overload of WTF::arePointingToEqualData that
does the right thing.

* Source/WTF/wtf/Ref.h:
* Source/WTF/wtf/UniqueRef.h:
    - Adds overloads.

* Source/WebCore/css/CSSCustomPropertyValue.cpp:
* Source/WebCore/css/CSSFontStyleRangeValue.cpp:
* Source/WebCore/css/CSSVariableReferenceValue.cpp:
* Source/WebCore/css/calc/CSSCalcTree.h:
* Source/WebCore/css/values/CSSValueAggregates.h:
* Source/WebCore/css/values/color/CSSDynamicRangeLimitMix.h:
* Source/WebCore/platform/graphics/FontCascadeDescription.h:
* Source/WebCore/platform/graphics/avfoundation/objc/CDMSessionAVContentKeySession.mm:
* Source/WebCore/platform/graphics/filters/FELighting.cpp:
* Source/WebCore/platform/graphics/filters/FilterOperations.cpp:
* Source/WebCore/platform/graphics/transforms/TransformOperations.cpp:
* Source/WebCore/rendering/style/StyleCursorImage.cpp:
* Source/WebCore/rendering/style/StyleMultiImage.cpp:
* Source/WebCore/style/StyleCustomProperty.cpp:
* Source/WebCore/style/StyleCustomProperty.h:
* Source/WebCore/svg/graphics/filters/SVGFilter.cpp:
    - Adopt new arePointingToEqualData overloads.

Canonical link: https://commits.webkit.org/296358@main
